src/types/nbt.rs

```diff
@@ -25,377 +25,291 @@ pub enum Nbt {
-            List::Int(ref value) => 4 * value.len(),
-            List::Long(ref value) => 8 * value.len(),
-            List::Float(ref value) => 4 * value.len(),
-            List::Double(ref value) => 8 * value.len(),
-            List::ByteArray(ref value) => 4 + value.len(),
-            List::String(ref value) => 2 + value.len(),
-            List::List(ref value) => {
-                5 + value.iter().map(|c| <List as Protocol>::proto_len(c)).sum()
-            }
-            List::Compound(ref value) => {
-                1 + value.iter().map(|c| {
-                    c.iter().map(|(name, nbt)| {
-                        2 + name.len() + <Nbt as Protocol>::proto_len(nbt)
-                    }).sum()
-                }).sum()
-            }
-            List::IntArray(ref value) => 4 + 4 * value.len(),
-        }
+impl Protocol for NbtFile {
+    type Clean = NbtFile;
+
+    fn proto_len(value: &NbtFile) -> usize {
+        value.len()
     }
 
-    fn proto_encode(value: &List, mut dst: &mut Write) -> io::Result<()> {
-        match value {
-            &List::Byte(ref xs) =>      try!(Nbt::write_list::<i8>(0x01, xs, dst)),
-            &List::Short(ref xs) =>     try!(Nbt::write_list::<i16>(0x02, xs, dst)),
-            &List::Int(ref xs) =>       try!(Nbt::write_list::<i32>(0x03, xs, dst)),
-            &List::Long(ref xs) =>      try!(Nbt::write_list::<i64>(0x04, xs, dst)),
-            &List::Float(ref xs) =>     try!(Nbt::write_list::<f32>(0x05, xs, dst)),
-            &List::Double(ref xs) =>    try!(Nbt::write_list::<f64>(0x06, xs, dst)),
-            &List::ByteArray(ref xs) => {
-                try!(dst.write_i8(0x07));
-                try!(dst.write_i32::<BigEndian>(xs.len() as i32));
-                for array in xs.iter() {
-                    try!(Nbt::write_i8_array(array, dst));
-                }
-            }
-            &List::String(ref xs) => {
-                try!(dst.write_i8(0x08));
-                try!(dst.write_i32::<BigEndian>(xs.len() as i32));
-                for value in xs.iter() {
-                    try!(Nbt::write_str(value.as_slice(), dst));
-                }
-            }
-            &List::List(ref xs) => try!(Nbt::write_list::<List>(0x09, xs, dst)),
-            &List::Compound(ref xs) => {
-                try!(dst.write_i8(0x0a));
-                try!(dst.write_i32::<BigEndian>(xs.len() as i32));
-                for compound in xs.iter() {
-                    try!(Nbt::write_compound(compound, dst));
-                }
-            }
-            &List::IntArray(ref xs) => {
-                try!(dst.write_i8(0x0b));
-                try!(dst.write_i32::<BigEndian>(xs.len() as i32));
-                for array in xs.iter() {
-                    try!(Nbt::write_i32_array(array, dst));
-                }
-            }
-        }
-        Ok(())
+    fn proto_encode(value: &NbtFile, dst: &mut io::Write) -> io::Result<()> {
+        value.write(&mut *dst)
```

Comment:
It gives an `Error: cannot borrow ... as mutable` without it. @toqueteos pointed out another way to fix it, which I'll use instead.